### PR TITLE
Reset password

### PR DESCRIPTION
**Pull request by romain-ortega**
Allow each user to reset their password with a dedicated page.
It works as follows:
1. Go to reset password page
2. Fill email address field
3. Click on the link received by email
4. Update the password

_Todo: destroy all reset password tokens which are link to a user after each new connection._

### CODE REVIEW

File: assets/app/reset-password/token/controller.js [hunk 1]
```diff
@@ -35,6 +35,7 @@ export default Ember.Controller.extend({
 					this.transitionToRoute('login');
 				},
 				(err) => {
+          console.log(err);
```
Comment (spacewoox) [inline assets/app/reset-password/token/controller.js]: console log

File: assets/app/reset-password/template.hbs [hunk 2]
```diff
@@ -1 +1,2 @@
 {{outlet}}
+
```
Comment (spacewoox) [inline assets/app/reset-password/template.hbs]: white space

File: assets/tests/unit/reset-password/token/route-test.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General
+ * Public License
+ * along with this program.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+
```
Comment (spacewoox) [inline assets/tests/unit/reset-password/token/route-test.js]: useless whitespace

File: tests/api/reset-password.test.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General
+ * Public License
+ * along with this program.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+
```
Comment (spacewoox) [inline tests/api/reset-password.test.js]: useless whitespace

File: api/controllers/Reset-passwordController.js [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+      });
+    })
+    .then(() => {
+      return ConfigService.get('host');
+    })
+    // send him reset password link
+    .then((conf) => {
+      let subject = 'Nanocloud - Reset your password';
+      let message = "Hello,<br>" +
+        "We got a request to reset your password.<br>" +
```
Comment (dynamiccast) [inline api/controllers/Reset-passwordController.js]: got -> received

File: api/controllers/Reset-passwordController.js [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        "<a href='"+conf.host+"/#/reset-password/"+token+"'>" +
+        "Reset my password</a><br><br><i>" +
+        "If you ignore thie message your password won't be changed.</i>";
```
Comment (dynamiccast) [inline api/controllers/Reset-passwordController.js]: thie -> this

File: tests/api/reset-password.test.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+          .expect(nano.jsonApiSchema(expectedSchema))
+          .then(() => {
+            // token has been added to 'reset-password' table
+            return (nano.request(sails.hooks.http.app)
+              .get('/api/reset-passwords')
+              .set(nano.adminLogin())
+              .expect(200)
+              .expect(nano.jsonApiSchema(expectedSchema))
+              .expect((res) => {
+                console.log(res.body.data[0]);
```
Comment (spacewoox) [inline tests/api/reset-password.test.js]: log message

File: tests/api/reset-password.test.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+                  .to.equal('admin@nanocloud.');
+              })
+            );
+          });
+        })
+        .then(() => {
+          return done(); 
+        })
+        .catch((err) => {
+          console.log("failure");
```
Comment (spacewoox) [inline tests/api/reset-password.test.js]: log message

File: api/controllers/Reset-passwordController.js [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    var token    = req.params.id;
+    var dataReq  = req.body.data.attributes;
+
+    // find user
+    ResetPassword.findOne({
+      "id": token
+    })
+    .then((tokenFound) => {
+      if (!tokenFound) {
+        throw new Error('Token has expired');
```
Comment (dynamiccast) [inline api/controllers/Reset-passwordController.js]: Token has expired -> Token expired

File: api/controllers/Reset-passwordController.js [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+      }
+      token = tokenFound;
+      return User.findOne({
+        "email": token.email
+      });
+    })
+    // update his password
+    .then((user) => {
+      if (!user) {
+        throw new Error('No user has been found');
```
Comment (dynamiccast) [inline api/controllers/Reset-passwordController.js]: No user has been found -> No user found

File: api/controllers/Reset-passwordController.js [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+      return ResetPassword.destroy({
+        id: token.id
+      });
+    })
+    // return response
+    .then(() => {
+      return res.ok({});
+    })
+    .catch((err) => {
+      if (err.message === 'No user has been found') {
```
Comment (dynamiccast) [inline api/controllers/Reset-passwordController.js]: Match updated error messages above

File: assets/app/router.js [hunk 12]
```diff
@@ -49,7 +49,7 @@ Router.map(function() {
       this.route('machine', { path: '/:machine_id' });
     });
     this.route('apps', function() {
-      this.route('app', { path: '/:app_id' });
+      this.route('app', { path: '/:app_id' }, function() {});
```
Comment (dynamiccast) [inline assets/app/router.js]: Why ?

File: tests/api/reset-password.test.js (hunk 7 above)
Comment (dynamiccast) [inline tests/api/reset-password.test.js]: Remove console.log please.

File: tests/api/reset-password.test.js [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+            // token has been added to 'reset-password' table
+            return (nano.request(sails.hooks.http.app)
+              .get('/api/reset-passwords')
+              .set(nano.adminLogin())
+              .expect(200)
+              .expect(nano.jsonApiSchema(expectedSchema))
+              .expect((res) => {
+                console.log(res.body.data[0]);
+                expect(res.body.data[0].attributes["email"])
+                  .to.equal('admin@nanocloud.');
```
Comment (dynamiccast) [inline tests/api/reset-password.test.js]: This email is invalid. Why would it work ?

File: tests/api/reset-password.test.js [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+              })
+            );
+          });
+        })
+        .then(() => {
+          return done(); 
+        })
+        .catch((err) => {
```
Comment (dynamiccast) [inline tests/api/reset-password.test.js]: Do not catch here. Just let the error bubble

**Comment by dynamiccast** [comment]
`/api/reset-passwords` is still under the `isAuthorized` policy.

**Comment by dynamiccast** [comment]
Not part of your PR but we need to fix this : `nanocloud/assets/app/reset-password/index/controller.js:38`. _Please check out you emails_ -> _Please check out your emails_

**Comment by dynamiccast** [comment]
We will need live frontend validation on password. It must follow the same requirement as in `user/new`.

**Pull request merged**